### DISCUSSION TITLE
K8S operator for ValKey?

### DISCUSSION THREAD

**Discussion by arpan57** [valkey-io/valkey#746] [Q&A]
Hello community, 

Do we have an operator for ValKey? We are excited to try ValKey as a redis replacement on a Kubernetes cluster. I briefly tried using existing [Redis operator](https://github.com/OT-CONTAINER-KIT/redis-operator)  without success.
Please suggest if you have successfully used same or any other operator with ValKey.  

Many thanks.

**Comment by zuiderkwast** [comment]
We (Ericsson) are using an internal fork of this one: https://github.com/AmadeusITGroup/Redis-Operator (The open source one is very old and not very good.) Ours is not really ready to be open sourced, for various reasons. We would really like to see an official open source operator for Valkey Cluster though.

IBM has one for cluster as well: https://github.com/IBM/operator-for-redis-cluster. I don't know of it's an updated fork of the Amadeus one. Also, I don't know the status.

I believe you need different functionality depending on whether you're running cluster or not.

I heard Canonical is working on an operator, but it seems very unclear if it's for cluster or not and if it's even for K8s. From what I heard, it can run on K8s or OpenStack or other platforms as well, so I 
don't really know how that plays with K8s. (I wonder if it's using K8s Operator SDK.

**Reply by arpan57** [reply]
Thank you @zuiderkwast. Appreciate your comment. 
We would like to run both ValKey Clusters and ValKey instances depending on the customer requirements.  It would be great to see if there are any trusted and working operators for ValKey. I briefly tried using existing [Redis operator](https://github.com/OT-CONTAINER-KIT/redis-operator) to build a ValKey cluster but it didn't work. 
Thanks.

**Reply by zuiderkwast** [reply]
Great. Let me know if you have any success.

For the record, we'd need at least the following features for cluster:

* Orchestrate scaling, i.e. add/remove nodes and migrate slots between them.
* Rolling upgrades without service impact.
* Anti-affinity handling, i.e. don't put primary and its replicas on the same worker.

**Comment by dmolik** [comment]
Hi I started work on an operator: https://github.com/hyperspike/valkey-operator/ I'm still working on finishing up the initial PoC but would love feedback.

I'll be sure to incorporate @zuiderkwast feature requests in the near future.

At the moment I'm working on status updates, upgrading/scaling, and events

**Reply by dmolik** [reply]
okay it takes about 50 seconds to bring up a 3 node cluster locally, I believe that includes cached docker images, anything faster will require changes to the bitnami docker image

**Reply by dmolik** [reply]
now have a cluster ready status:
![image](https://github.com/user-attachments/assets/290ec94e-679c-407b-ad82-5b8b7877e8c7)

**Reply by arpan57** [reply]
Thanks for the initiative @dmolik . 
I cloned the operator and tried applying the config/sample. I can see the similar output showed in your scrneeshot.   
However, I do not see any pods.

**Reply by dmolik** [reply]
This is probably some up with the statefulset, could you create an issue in repo, or post the output of `kubectl describe sts`

**Reply by arpan57** [reply]
```
❯ k get valkeys.hyperspike.io
NAME            READY   AGE   NODES   REPLICAS
valkey-sample           2s    3       0
❯ k get pods
No resources found in default namespace.
❯ k get sts
No resources found in default namespace.
```
Btw - i am running it in a minikube in a default namespace for now.

**Reply by dmolik** [reply]
So far, I've been doing all of my testing using a 3 node minikube install. What's the log output of the controller?
`kubectl logs -n valkey-operator-system -l control-plane=controller-manager`

**Reply by dmolik** [reply]
there is also a `make` target for creating a local cluster; `make minikube` which I make no guarantees about working with anything other than a Linux workstation

**Reply by arpan57** [reply]
Thanks for quick response. I am on a mac. I tried `make minikube` It created a new clustercontext north. 
I do not see the valkey-operator-system namespace in either of the clusters (north and minikube) . May be that's the reason I might not have a controller which would act on CRD resources. If three node minikube then, will make minikube not create another instance like north?

**Reply by dmolik** [reply]
that could be it.
`kubectl get ns`? 

It could also be that the container image was not published publically

**Reply by dmolik** [reply]
okay that took an embarrassing amount of time but the containers are now plubic

**Reply by arpan57** [reply]
```
Switched to context "north".
❯ k get ns
NAME              STATUS   AGE
cilium-spire      Active   33m
default           Active   34m
ingress-nginx     Active   26m
kube-node-lease   Active   34m
kube-public       Active   34m
kube-system       Active   34m
```


```
❯ kubectx minikube
Switched to context "minikube".
❯ k get ns
NAME                   STATUS   AGE
default                Active   2d1h
kube-node-lease        Active   2d1h
kube-public            Active   2d1h
kube-system            Active   2d1h
kubernetes-dashboard   Active   2d1h
minikube-pv-hack       Active   21m
```

**Reply by arpan57** [reply]
I think I will start a discussion on operator's git/issues. :-)

**Reply by dmolik** [reply]
weird a namespace should be created. Try this:

`kubectl apply -f https://raw.githubusercontent.com/hyperspike/valkey-operator/main/dist/install.yaml`

**Reply by arpan57** [reply]
What I did was, stopped minikube. 
$ make minikube
This created the name space (valkey-operator-system) and also the 3 pods :-) All the pods are stuck with FailedScheduling state
At least some success :-) 

```
Events:
  Type     Reason            Age                  From               Message
  ----     ------            ----                 ----               -------
  Warning  FailedScheduling  2m7s (x2 over 2m8s)  default-scheduler  0/4 nodes are available: pod has unbound immediate PersistentVolumeClaims. preemption: 0/4 nodes are available: 4 Preemption is not helpful for scheduling.
  Warning  FailedScheduling  118s                 default-scheduler  0/5 nodes are available: pod has unbound immediate PersistentVolumeClaims. preemption: 0/5 nodes are available: 5 Preemption is not helpful for scheduling.
  Warning  FailedScheduling  1s (x2 over 97s)     default-scheduler  0/6 nodes are available: pod has unbound immediate PersistentVolumeClaims. preemption: 0/6 nodes are available: 6 Preemption is not helpful for scheduling.

```

**Comment by hoterm** [comment]
@arpan57 FYI - you can join the Slack channel for Valkey community if you want, could be useful

**Reply by arpan57** [reply]
Thanks @hoterm  - Couldn't find it on https://valkey.io/connect/ 

Can you please point me me Valkey related slack channel? 

Thank you.

**Reply by karuppiah7890** [reply]
Here you go! https://valkey-oss-developer.slack.com/join/shared_invite/zt-2nxs51chx-EB9hu9Qdch3GMfRcztTSkQ

I think it's not mentioned on the Valkey official website. I noticed and found it on @madolson's LinkedIn post here - https://www.linkedin.com/posts/madelyn-olson-valkey_github-valkey-iovalkey-a-flexible-distributed-activity-7249156680221929472-_IIj

**Reply by karuppiah7890** [reply]
Created an issue for fixing it in the website - https://github.com/valkey-io/valkey-io.github.io/issues/198

**Reply by arpan57** [reply]
Thanks @karuppiah7890 .

**Comment by pierreozoux** [comment]
For the curious like me, the summary is:
- Amadeus operator is dead, not maintained any more
- IBM version, same
- Ericson version, internal fork of Amadeus
- OT container kit [issue to support valkey](https://github.com/OT-CONTAINER-KIT/redis-operator/issues/905) (a year ago Percona was looking to contribute to it)
- bitnami chart is not an operator (: ( [and not really open source any way](https://github.com/bitnami/containers/issues/73236#issuecomment-2413133047) )
- https://github.com/hyperspike/valkey-operator looks like the only viable option 😃 thanks @dmolik for working on this!

**Comment by arpan57** [comment]
I have been working on adapting OT container kit operator for Valkey. I have patched it for running Valkey in cluster mode. So, far it's been working well in Dev environments.

I tried Valkey operator in last year July/Aug and it was pretty new may be one or two months old with only one contributor. I am sure it will get ready soon.  When I evaluated operators I had to decide to patch the OT container kit one for Valkey and move ahead.

Thanks for the summary.

**Reply by zuiderkwast** [reply]
Can it do rolling upgrades in the way needed to keep Valkey Cluster up? I mean add new upgraded nodes as replicas, wait for them to sync with the primary, then switch over so the old primaries become a replicas, then take down the old primaries? This is one of the important non-trivial things we would need to have.

**Reply by sandeepkunusoth** [reply]
@arpan57 can u pease answer @zuiderkwast  question

**Comment by pierreozoux** [comment]
And there is a new valkey operator :)

https://github.com/SAP/valkey-operator

(from their repo:
The controller contained in this repository under the hood uses the [bitnami valkey chart](https://github.com/bitnami/charts/tree/main/bitnami/valkey) to install valkey in the cluster. As a consequence of this fact, the following topologies are supported:

I'd say then as bitnami is not really open source, this operator neither)
)

**Reply by zuiderkwast** [reply]
Good to know. I read this on the project:

> Sharding (valkey-cluster) scenarios are not supported.

**Comment by ovaiskhansc** [comment]
Late to this discussion. We at Snap are using a fork of IBM's redis operator: https://github.com/IBM/operator-for-redis-cluster. We would be happy to collaborate with others on a ValKey Kubernetes Operator

**Reply by vadlamak** [reply]
We have built an operator at Geico 
Based on https://github.com/OT-CONTAINER-KIT/redis-operator

We would like to contribute to Valkey Operator as well.

**Comment by chideat** [comment]
Hi everyone,

I've developed a new Valkey Operator for Kubernetes and wanted to share it with the community. You can find it here: [https://github.com/chideat/valkey-operator](https://github.com/chideat/valkey-operator)

It's designed for production use and already supports a solid set of features:
* Architectures: Standalone, Sentinel, and Cluster modes.
* Operations: Online scaling (up/down) and graceful version upgrades.
* Security & Networking: Valkey ACL, NodePort/LoadBalancer access, IPv4/IPv6.
* Scheduling: nodeSelector, tolerations, and affinity are all supported.
* High Availability: Built with production HA in mind.

I would greatly appreciate any feedback, issues, or contributions. Hope you find it useful!
